### ISSUE TITLE
resin device register only does the old long UUIDs

### ISSUE THREAD

**Issue by imrehg** [resin-io/resin-cli#473]
With resinOS 2.0, the device UUIDs became half the length (~31~ 32 chars down from 62 characters). Running `resin device register AppName` only able to do the original 62 character version. Not sure if it's a problem, just highlight it, because this way images freshly downloaded from the Dashboard and images configured with the CLI will have different behaviour for 2.0.

What would be the right default behaviour? Or this just doesn't have any real consequences?

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_20dt)

**Comment by imrehg** [comment]
Actually, using the short UUID, `resin config generate` fails too, so can't use CLI for generating config for devices provisioned by images downloaded from the dashboard:

```
$ export UUID=b8c8f07468c6cfaa9aacf0c06b5e22a
$ resin config generate --device ${UUID}
? Network Connection ethernet
? Check for updates every X minutes 1
Validation: uuid invalid input
```

**Comment by emirotin** [comment]
This should be easy to fix but this is a process fail: now is the first time I hear about some changes to the UUID generation.

**Comment by imrehg** [comment]
Likely relevant: https://github.com/resin-os/meta-resin/commit/4e537949ce9131f2e47d8e24ac5e366e31e38dd6

**Comment by pimterry** [comment]
Just seen this - turns out we actually do have an outstanding issue with ResinOS 2 then.

The config generation issue is in [resin-device-config](https://github.com/resin-io-modules/resin-device-config/blob/a6721d65ab51a9436f13b0ee63d1c41c8c1220af/lib/schema.coffee#L146) which is supposed to validate the generated device config, and has strong opinions on what UUIDs should look like. I'm thinking to to change this to validate it's just a _minimum_ of 16 bytes. Does that sound reasonable?

I'm not sure what to do about `resin device register` coming back with 62 byte UUIDs. Will that break things?  That comes from [resin-register-device](https://github.com/resin-io-modules/resin-register-device/blob/master/lib/register.coffee#L28-L51), which currently generates UUIDs and device keys with the same method. More problematically, `resin device register` doesn't know which OS it's going to be using yet, so can't easily switch between the two cases. Not sure what we should do there to handle that...

**Comment by imrehg** [comment]
thanks @pimterry. Yeah, the "is it a 1.x or 2.x device?" is a curious thing, because it will be a mixture later anyways (e.g. devices upgraded from 1.x to 2.x will be unlikely to get their UUIDs truncated, I guess). So I think generating long UUIDs shouldn't be a too big of a problem from the SDK's point of view, but the validation is. 

A note, that currently `resin device register MyApp  --uuid <something>` can create any kind of UUID (for example I set one just to `test`, and it was accepted. So there might be good to have some validation there too, so that the CLI is sort of self-consistent?

**Comment by pimterry** [comment]
There's now a fix for the main issue here: https://github.com/resin-io-modules/resin-device-config/pull/44.

I've pulled out a separate issue for 'make UUID validation consistent' at https://github.com/resin-io/resin-cli/issues/481, because it's not urgent, and needs more investigation as a wider issue. Once that resin-device-config PR is merged this issue can be closed.

**Comment by pimterry** [comment]
@imrehg Where did you get the UUID you used above from? @emirotin has pointed out there's some bytes/characters confusion here. Looking at the meta-resin change above more closely, we're moving to a 16 _byte_ UUID, which should lead to a 32 character UUID string, not 16 characters. Yours is 31 though. Any idea why?

**Comment by imrehg** [comment]
@pimterry sorry, it was an oversight. (yeah, 32 character UUID instead of 31 I've said originally).

I've got the UUID from the dashboard, copied from a 2.0.0 device.

**Comment by pimterry** [comment]
Thanks @imrehg! This has now been fixed in `resin-device-config@4.0.1`. This will work in the next release, once https://github.com/resin-io/resin-cli/pull/487 (which bumps us to `^4.0.0`) has been merged.

**Comment by imrehg** [comment]
@pimterry any update on this one? looks like still an issue on `6.1.0` (e.g. try with `resin config generate --device ${UUID}` where UUID is of a 2.x device.

**Comment by pimterry** [comment]
This fix is ready in `resin-device-config`, but only on v4, which has a breaking change from @Page- that needs the corresponding changes started in #487 to be finished before we can upgrade in the CLI.

@Page- are you going to get a chance to finish that PR any time soon? Looks like it's blocked on supporting pre-2.x devices at the mo. If not, we're going to have to find a nice way to do a patch release off the old v3 code for `resin-device-config` to get this bug fixed.

**Comment by pimterry** [comment]
The `config generate` issue is now fixed and released in 6.7.0, so you can once again generate config for ResinOS 2.x device UUIDs.

The `device register` issue, where it generates 62 char UUIDs by default, I'm not planning to fix - I'm not aware of any reason why you would _need_ to generate a shorter UUID for yourself, the current 62 char format works everywhere, and it's awkward and unnecessary to prompt every user for which format they'd prefer.

If there is any good reason we need that behaviour I would happily reopen this though, and in the meantime if anybody would simply prefer that behaviour for some reason, you can easily create it for yourself with `resin device register --uuid $(openssl rand -hex 16)`
